File: ios/CourtSearchKit/Sources/CourtSearchKit/SearchEngine.swift
import Foundation
import SwiftProtobuf


/// A failure the engine reported through its MobileResponse envelope, or an
/// envelope that carried neither payload nor error.
public struct EngineError: Error, CustomStringConvertible {
    public let operation: String
    public let code: Int
    public let message: String
    public var description: String { "\(operation): [\(code)] \(message)" }
}

/// Typed calls over the engine's byte ABI. Every call blocks until the engine
/// answers, so callers keep it off the main thread (CourtIndex is an actor).
final class SearchEngine {
    private(set) var handle: UInt64

    init(open request: Ai_Protomolt_Search_Mobile_V1_MobileOpenRequest, create: Bool) throws {
        let opened: Ai_Protomolt_Search_Mobile_V1_MobileOpenResponse = try Self.decode(
            "open", ProtomoltSearchMobile.open(try request.serializedData(), create: create))
        guard opened.noEgress else {
            throw EngineError(operation: "open", code: -1, message: "runtime did not certify no-egress")
        }
        handle = opened.handle
    }

    func planIndex(_ request: Ai_Protomolt_Search_V1_PlanIndexRequest) throws -> Ai_Protomolt_Search_V1_PlanIndexResponse {
        try Self.decode("planIndex", ProtomoltSearchMobile.planIndex(handle: handle, request: try request.serializedData()))
    }

    func ingestMapped(_ batch: Ai_Protomolt_Search_Mobile_V1_MobileIngestMappedBatch) throws -> Ai_Protomolt_Search_V1_IngestMappedResponse {
        try Self.decode("ingestMapped", ProtomoltSearchMobile.ingestMapped(handle: handle, request: try batch.serializedData()))
    }

    func query(_ request: Ai_Protomolt_Search_V1_QueryRequest) throws -> Ai_Protomolt_Search_V1_QueryResponse {
        try Self.decode("query", ProtomoltSearchMobile.query(handle: handle, request: try request.serializedData()))
    }

    func flush() throws {
        let _: Ai_Protomolt_Search_Mobile_V1_MobileFlushResponse = try Self.decode(
            "flush", ProtomoltSearchMobile.flush(handle: handle))
    }

    func close() throws {
        let _: Ai_Protomolt_Search_Mobile_V1_MobileCloseResponse = try Self.decode(
            "close", ProtomoltSearchMobile.close(handle: handle))
    }

    private static func decode<T: SwiftProtobuf.Message>(_ operation: String, _ bytes: Data) throws -> T {
        let envelope = try Ai_Protomolt_Search_Mobile_V1_MobileResponse(serializedBytes: bytes)
        switch envelope.outcome {
        case .payload(let payload):
            return try T(serializedBytes: payload)
        case .error(let error):
            throw EngineError(operation: operation, code: error.code.rawValue, message: error.message)
        case nil:
            throw EngineError(operation: operation, code: -1, message: "empty response envelope")
        }
    }
}

File: android/app/src/main/java/ai/pipestream/samples/courtsearch/SearchEngine.kt
package ai.pipestream.samples.courtsearch

import ai.pipestream.search.mobile.ProtomoltSearch
import ai.protomolt.search.mobile.v1.Mobile
import ai.protomolt.search.v1.Search
import com.google.protobuf.ByteString

/** A failure the engine reported through its MobileResponse envelope. */
class EngineException(val operation: String, val code: Int, message: String) :
    Exception("$operation: [$code] $message")

/**
 * Typed calls over the engine's byte ABI. Every call blocks until the engine
 * answers, so callers keep it off the main thread.
 */
class SearchEngine(request: Mobile.MobileOpenRequest, create: Boolean) : AutoCloseable {
    private val handle: Long

    init {
        val opened = Mobile.MobileOpenResponse.parseFrom(
            payload("open", ProtomoltSearch.nativeOpen(request.toByteArray(), create)))
        if (!opened.noEgress) throw EngineException("open", -1, "runtime did not certify no-egress")
        handle = opened.handle
    }

    fun planIndex(request: Search.PlanIndexRequest): Search.PlanIndexResponse =
        Search.PlanIndexResponse.parseFrom(
            payload("planIndex", ProtomoltSearch.nativePlanIndex(handle, request.toByteArray())))

    fun ingestMapped(batch: Mobile.MobileIngestMappedBatch): Search.IngestMappedResponse =
        Search.IngestMappedResponse.parseFrom(
            payload("ingestMapped", ProtomoltSearch.nativeIngestMapped(handle, batch.toByteArray())))

    fun query(request: Search.QueryRequest): Search.QueryResponse =
        Search.QueryResponse.parseFrom(
            payload("query", ProtomoltSearch.nativeQuery(handle, request.toByteArray())))

    fun flush() {
        payload("flush", ProtomoltSearch.nativeFlush(handle))
    }

    override fun close() {
        payload("close", ProtomoltSearch.nativeClose(handle))
    }

    private fun payload(operation: String, bytes: ByteArray): ByteString {
        val envelope = Mobile.MobileResponse.parseFrom(bytes)
        return when (envelope.outcomeCase) {
            Mobile.MobileResponse.OutcomeCase.PAYLOAD -> envelope.payload
            Mobile.MobileResponse.OutcomeCase.ERROR ->
                throw EngineException(operation, envelope.error.codeValue, envelope.error.message)
            else -> throw EngineException(operation, -1, "empty response envelope")
        }
    }
}
